src/prefect/engine/task_runner.py

```diff
@@ -152,12 +152,12 @@ def initialize_run(  # type: ignore
                 "task_loop_count": state.cached_inputs.pop(  # type: ignore
                     "_loop_count"
                 )  # type: ignore
-                .to_result()
+                .to_result() # todo: here
                 .value,
                 "task_loop_result": state.cached_inputs.pop(  # type: ignore
                     "_loop_result"
                 )  # type: ignore
-                .to_result()
+                .to_result() # todo: here
```

Comment:
Context here:

- this data needs to be stored on the `state` object for the task so that it can be persisted and retrieved
- because of this, this data needs a formal `Result` wrapper (otherwise they will be serialized to `None` / `NoResult`).  `JSONResult` is appropriate here because they are simple objects

These `to_result` calls are then basic "write" calls with the new interface, no filename templating needed.  I would expect the `write` call for `JSONResult` to store the output of `json.dumps` as an attribute of the class.  Or conversely, we can "write" the results within the serialization interface and not even need to explicitly call write here.